File: lang/php/patterns/structural/flyweight.php
<?php

interface Foo
{
  public function show();
}

class Bar implements Foo
{
  protected $value;

  public __construct(int $value)
  {
    $this->value = $value;
  }

  public function show()
  {
    return $this->value;
  }
}

class FooFactory
{
  protected $cache = [];

  public function get(int $value)
  {
    if (!array_key_exists($value, $this->cache))
    {
      $this->cache[$value] = new Bar($value);
    }

    return $this->cache[$value];
  }
}

File: lang/c++/patterns/flyweight.cpp
#include <iostream>
#include <vector>

class Foo
{
public:
	int Value;

	Foo(int v) :Value(v) {}
};

class FooFactory
{
private:
	std::vector<Foo*> cache;

public:
	Foo* Get(int v)
	{
		for (auto i : cache)
			if (i->Value == v)
				return i;

		Foo foo(v);

		cache.push_back(&foo);

		return &foo;
	}
};

int main()
{
	FooFactory factory;

	std::cout << factory.Get(1) << std::endl;
	std::cout << factory.Get(1) << std::endl;
	std::cout << factory.Get(2) << std::endl;
	std::cout << factory.Get(2) << std::endl;
	std::cout << factory.Get(3) << std::endl;
	
	return 0;
}